### PR TITLE
gameEvents: created `ANNOUNCE_UNO` event

### PR DESCRIPTION
**Pull request by sethdivyansh**
Fixes: #132

## Description
In short:
- Created `announced_uno` event.
- Added `runningEvents` with `vulnerableToUNO` and `hasAnnouncedUNO` subfields in engine.ts.
- Created `getThrowableCards()` in eventHandlerUtils.
- Handled corner cases related to the announced_uno event.
- Created tests for the announced_uno event and its corner cases.

The plan this PR followed to handle `ANNOUNCE_UNO` event:
- Players can announce UNO only if:
  - It's their turn and they have 2 cards, with at least one throwable card.
  - OR before the other players catch him not saying UNO.
- If a player fails to say "UNO" and isn't caught before the next player plays, they are considered to have announced UNO.
- Added `runningEvents` field in GameEngine with sub-fields `vulnerableToUNO` and `hasAnnouncedUNO`.
- Checked if the new player can announce UNO on each turn change.
- if the player is in position to say UNO, set `vulnerableToUNO` to this player.
- if the player announces UNO, it will set `hasAnnounedUNO` to this player, and set `vulnerableToUNO` to null.
- set `vulnerableToUNO` to null when the next player plays his turn.
## How to Test
I have created some tests in `events.test.ts`

## Checklist
- [x] I have tested these changes locally.
- [x] I have reviewed the code and ensured it follows the project's coding guidelines.
- [ ] I have updated the documentation, if necessary.
- [ ] I have assigned reviewers to this pull request.

## Screenshots (if applicable)
[If your changes include any visual updates, provide screenshots here.]

### CODE REVIEW

File: backend/src/uno-game-engine/engine.ts [hunk 1]
```diff
@@ -10,6 +11,7 @@ export class GameEngine {
     thrownCards: UNOCard[];
     players: Player[];
     currentPlayerIndex: number;
+    prevPlayerIndex: number;
```
Comment (kuv2707) [inline backend/src/uno-game-engine/engine.ts:14]: We can compute the previous player given the current player and the direction. So we can get rid of this field

File: backend/src/uno-game-engine/events/announceUno.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    if (canThrow.type === 'ERROR') {
+        return canThrow;
+    }
+    // The moment a player has just one card they must yell “UNO!”.
+    if (player.cards.length !== 1 || player.hasAnnouncedUNO === true) {
+        return { type: 'ERROR', message: 'Cannot announce UNO' };
+    }
+
+    return { type: 'SUCCESS', message: 'Can announce UNO' };
+}
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/announceUno.ts:21]: Hmm, thinking about the UX here, in reality, a player shouts UNO as soon as they throw the second card, but we'd have to handle it differently in code. 

I got to know about how other apps implement this - they allow the UNO button to be pressed only when - 
- it is the player's turn
-  AND they have two cards one of which can be thrown to react the 1-card state. 

So while handling the event, we need to assert that the player has two cards, and its their turn and they have a throwable card.

You'd need to define a utility function in `eventHandlerUtils` which checks if the current player has a throwable card based on the last thrown card.


Also to enable/disable the button, we'd need to check the same conditions on turn change too (we'll handle that later, lets just have the button always enabled for each player now)

File: backend/src/types.ts [hunk 3]
```diff
@@ -35,6 +35,7 @@ export type UNOCard = {
 export type Player = {
     id: string;
     cards: UNOCard[];
+    hasAnnouncedUNO: boolean;
```
Comment (kuv2707) [inline backend/src/types.ts:38]: It would not be prudent to have this field on the Player itself, since the game is quite fast paced and things could get complicated while syncing player state with the clients. 

We should rather make accomodations for these temporary game states in the game object itself. This will need some planning to decide where to put

**Review by kuv2707 — COMMENTED** [review]
Huh, this issue seems more complicated than anticipated.

Most importantly, we need to find a structurally sound way to store temporary game states like player_announced_uno.

Here's the plan - 
- We create a field `runningEvents` on the `GameEngine`, which contains sub-fields `vulnerableToUNO` and `vulnerableDuration`.
- On each turn change, we check if the new player is in the position to announce uno (as described in a previous comment)
- if yes, we set `vulnerableToUNO` to this player, and `vulnerableDuration` to the number of players (denoting one round)
- When the player presses the button meaning that `ANNOUNCE_UNO` is triggered, we unset `vulnerableToUNO` (only if the vulnerable player triggers the event obviously)
- After the current player's turn, for the next one round, other players have the option to `CALL_OUT_UNO` (a separate event), which then allots the vulnerable player 7-8 cards (i dont recall) as long as the vulnerableDuration is not 0.
- Each time we move to the next player, we also decrement `vulnerableUNO` field if it is not null.

I know it is complicated, so do ask some follow-ups / suggest improvements to this

File: backend/src/uno-game-engine/events/announceUno.ts (hunk 2 above)
Comment (sethdivyansh) [reply backend/src/uno-game-engine/events/announceUno.ts:21]: I was also thinking of allowing the player to announce UNO when they have two cards, but then I read the rules, which say
image

However, I will allow them to shout UNO when they have two cards remaining and it’s their turn.

File: backend/src/uno-game-engine/engine.ts (hunk 1 above)
Comment (sethdivyansh) [reply backend/src/uno-game-engine/engine.ts:14]: like this or is there anyother better way of finding `prePlayerIndex`

```javascript
const prevPlayerIndex =
        (game.players.length + currentPlayerIndex - game.direction) %
        game.players.length;
```

File: backend/src/uno-game-engine/events/announceUno.ts (hunk 2 above)
Comment (kuv2707) [reply backend/src/uno-game-engine/events/announceUno.ts:21]: It's hard to replicate the rule in a website because of network lags etc.

File: backend/src/uno-game-engine/engine.ts (hunk 1 above)
Comment (kuv2707) [reply backend/src/uno-game-engine/engine.ts:14]: This is it

**Comment by kuv2707** [comment]
In that case, we'll go with the rules. It'll be easier to implement too

**Comment by sethdivyansh** [comment]
@kuv2707
I have made the changes and followed your structure with some modifications. Please see the updated structure in the PR description.

File: backend/src/types.ts [hunk 4]
```diff
@@ -47,16 +53,13 @@ export enum GameEventTypes {
...
     LEAVE_GAME = 'LEAVE_GAME',
+    ANNOUNCE_UNO = 'ANNOUNCE_UNO',
     STATE_SYNC = 'STATE_SYNC',
 }
-
 export type GameEvent =
     | {
           type: GameEventTypes.DRAW_CARD;
           playerId: string;
-          data: {
```
Comment (kuv2707) [inline backend/src/types.ts:57]: Make it data?:null so that the overall GameEvent type always has a data property (this is just to satisfy TS)

File: backend/src/uno-game-engine/events/eventHandlerUtils.ts [hunk 5]
```diff
@@ -11,6 +11,21 @@ export function getPlayerCard(player: Player, cardId: string) {
...
 
+export function getThrowableCards(game: GameEngine, player: Player) {
+    // get the cards that the player can throw
+    const { lastThrownCard } = game;
+    const throwableCards = player.cards.filter((card) => {
+        return (
+            !lastThrownCard ||
+            card.color === lastThrownCard.color ||
+            card.value === lastThrownCard.value ||
+            card.color === 'wild'
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/eventHandlerUtils.ts:22]: make it card.type to make semantic sense ( i suppose both are set to `wild`)

File: backend/src/uno-game-engine/events/announceUno.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    const throwableCards = getThrowableCards(game, player);
+    if (
+        player.cards.length > 2 ||
+        (player.cards.length === 2 && !throwableCards)
+    ) {
+        return { type: 'ERROR', message: 'Cannot announce UNO' };
+    }
+
+    return { type: 'SUCCESS', message: 'Can announce UNO' };
+}
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/announceUno.ts:19]: Seems like you missed checking whether the player actually has a turn now.

Comment (kuv2707) [reply backend/src/uno-game-engine/events/announceUno.ts:19]: there is a utility fn in eventHandlerUtils for that

File: backend/src/uno-game-engine/events/drawCard.ts [hunk 7]
```diff
@@ -22,6 +22,11 @@ export function drawCard(game: GameEngine, event: GameEvent): EventResult {
         return drawResult;
     }
 
+    if (player.cards.length === 2) {
+        game.runningEvents.hasAnnouncedUNO = null;
+        game.runningEvents.vulnerableToUNO = null;
+    }
+
```
Comment (kuv2707) [inline backend/src/uno-game-engine/events/drawCard.ts:29]: I'm having trouble understanding why this is required

Comment (kuv2707) [reply backend/src/uno-game-engine/events/drawCard.ts:29]: hmm, got it after reading the tests. 
Any kind of action by the subsequent player, which includes both drawing card or throwing one, should reset the vulnerablePlayer.

File: backend/src/uno-game-engine/events/eventHandlerUtils.ts (hunk 5 above)
Comment (kuv2707) [reply backend/src/uno-game-engine/events/eventHandlerUtils.ts:22]: this is too small a change to consider right now though

**Review by kuv2707 — COMMENTED** [review]
Seems like we are allowing announce_uno when either the player has 2 cards and one of them is throwable, or he has only one card. This means we allow the player to announce uno even when its next player's turn now. I am a little worried it would lead to race conditions, but the idea seems exciting right now (and would add to the thrill), so we can integrate this for now.

Seems good enough to merge

**Comment by kuv2707** [comment]
Please address [this](https://github.com/shivansh-bhatnagar18/multiplayer-uno/pull/136#discussion_r1649395107) to fix the CI test.

**Comment by sethdivyansh** [comment]
> Seems like we are allowing announce_uno when either the player has 2 cards and one of them is throwable, or he has only one card. This means we allow the player to announce uno even when its next player's turn now(which wasn't originally what I had thought). I am a little worried it would lead to race conditions, but the idea seems exciting right now (and would add to the thrill), so we can integrate this for now and see if it works well.
> 
> Seems good enough to merge, once the minor remarks are addressed

Yes, you are  right. I have made it such that players can announce UNO even if it's next player's turn. As players can be caught not saying "UNO" before the next player has taken their turn.

**Comment by sethdivyansh** [comment]
Another idea: 
we may do it in this way:
give other players 4-5s (or more or less) to catch the unannounced player. If they fail, we'll remove 'vulnerableToUNO' from that player.

**Comment by kuv2707** [comment]
This seems good for now, we can optimize this code later if need arises.

**Comment by kuv2707** [comment]
Merged, thanks @sethdivyansh
